File: Whatsapp/OpenWA/sdk/java/src/main/java/com/rmyndharis/openwa/resources/GroupsResource.java
package com.rmyndharis.openwa.resources;

import static com.rmyndharis.openwa.http.Http.encodeSegment;

import com.rmyndharis.openwa.OpenWAClient;
import com.rmyndharis.openwa.http.HttpMethod;
import com.rmyndharis.openwa.model.CreateGroupRequest;
import com.rmyndharis.openwa.model.GroupDescriptionRequest;
import com.rmyndharis.openwa.model.GroupInfo;
import com.rmyndharis.openwa.model.GroupSettings;
import com.rmyndharis.openwa.model.GroupSubjectRequest;
import com.rmyndharis.openwa.model.GroupSummary;
import com.rmyndharis.openwa.model.InviteCodeResponse;
import com.rmyndharis.openwa.model.JoinGroupRequest;
import com.rmyndharis.openwa.model.JoinGroupResponse;
import com.rmyndharis.openwa.model.ListGroupsQuery;
import com.rmyndharis.openwa.model.ParticipantsRequest;
import com.rmyndharis.openwa.model.SuccessResult;
import java.util.List;

/** Groups resource — WhatsApp group management. */
public final class GroupsResource {
    private final OpenWAClient client;

    public GroupsResource(OpenWAClient client) {
        this.client = client;
    }

    /** List all groups for the session. */
    public List<GroupSummary> list(String sessionId) {
        return list(sessionId, null);
    }

    /** List all groups for the session, applying the given pagination query. */
    public List<GroupSummary> list(String sessionId, ListGroupsQuery query) {
        return client.requestList(
            HttpMethod.GET, "/api/sessions/" + encodeSegment(sessionId) + "/groups", query, null, GroupSummary.class);
    }

    /** Get detailed group info including the participant list. */
    public GroupInfo get(String sessionId, String groupId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId),
            null,
            null,
            GroupInfo.class);
    }

    /** Create a new group. */
    public GroupInfo create(String sessionId, CreateGroupRequest body) {
        return client.request(
            HttpMethod.POST, "/api/sessions/" + encodeSegment(sessionId) + "/groups", null, body, GroupInfo.class);
    }

    /** Add participants to a group. */
    public SuccessResult addParticipants(String sessionId, String groupId, List<String> participants) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/participants",
            null,
            new ParticipantsRequest(participants),
            SuccessResult.class);
    }

    /** Remove participants from a group. */
    public SuccessResult removeParticipants(String sessionId, String groupId, List<String> participants) {
        return client.request(
            HttpMethod.DELETE,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/participants",
            null,
            new ParticipantsRequest(participants),
            SuccessResult.class);
    }

    /** Promote participants to group admin. */
    public SuccessResult promoteParticipants(String sessionId, String groupId, List<String> participants) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/participants/promote",
            null,
            new ParticipantsRequest(participants),
            SuccessResult.class);
    }

    /** Demote participants from group admin. */
    public SuccessResult demoteParticipants(String sessionId, String groupId, List<String> participants) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/participants/demote",
            null,
            new ParticipantsRequest(participants),
            SuccessResult.class);
    }

    /** Update the group subject (name). */
    public SuccessResult setSubject(String sessionId, String groupId, String subject) {
        return client.request(
            HttpMethod.PUT,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/subject",
            null,
            new GroupSubjectRequest(subject),
            SuccessResult.class);
    }

    /** Join a group via an invite code. Returns the joined group id. */
    public JoinGroupResponse joinGroup(String sessionId, String inviteCode) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/join",
            null,
            new JoinGroupRequest(inviteCode),
            JoinGroupResponse.class);
    }

    /** Update the group description (empty string clears it). */
    public SuccessResult setDescription(String sessionId, String groupId, String description) {
        return client.request(
            HttpMethod.PUT,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/description",
            null,
            new GroupDescriptionRequest(description),
            SuccessResult.class);
    }

    /** Get the group settings (announce / locked / ephemeral timer). */
    public GroupSettings getGroupSettings(String sessionId, String groupId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/settings",
            null,
            null,
            GroupSettings.class);
    }

    /**
     * Update the group settings. At least one field of {@code settings} must be set; absent fields
     * stay untouched. Setting {@code ephemeralSeconds} returns HTTP 501 on the whatsapp-web.js engine.
     */
    public SuccessResult updateGroupSettings(String sessionId, String groupId, GroupSettings settings) {
        return client.request(
            HttpMethod.PUT,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/settings",
            null,
            settings,
            SuccessResult.class);
    }

    /** Leave a group. */
    public SuccessResult leave(String sessionId, String groupId) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/leave",
            null,
            null,
            SuccessResult.class);
    }

    /** Get the group invite code and link. */
    public InviteCodeResponse inviteCode(String sessionId, String groupId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/invite-code",
            null,
            null,
            InviteCodeResponse.class);
    }

    /** Revoke the current invite code and generate a new one. */
    public InviteCodeResponse revokeInviteCode(String sessionId, String groupId) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/groups/" + encodeSegment(groupId) + "/invite-code/revoke",
            null,
            null,
            InviteCodeResponse.class);
    }
}

File: Whatsapp/OpenWA/sdk/php/src/Resources/GroupsResource.php
<?php

declare(strict_types=1);

namespace OpenWA\Resources;

use OpenWA\Http\HttpExecutor;

/**
 * Groups resource — WhatsApp group management.
 *
 * Backed by src/modules/group/group.controller.ts.
 */
class GroupsResource
{
    private HttpExecutor $http;

    public function __construct(HttpExecutor $http)
    {
        $this->http = $http;
    }

    /**
     * @param array<string,mixed> $query
     * @return array<int,array<string,mixed>>
     */
    public function list(string $sessionId, array $query = []): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups", $query) ?? [];
    }

    /** @return array<string,mixed> */
    public function get(string $sessionId, string $groupId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}");
    }

    /**
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    public function create(string $sessionId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups", [], $body);
    }

    /**
     * Join a group via an invite code (the token from a chat.whatsapp.com/<code>
     * link). Returns {success: true, groupId}.
     *
     * @return array<string,mixed>
     */
    public function joinGroup(string $sessionId, string $inviteCode): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/join", [], ['inviteCode' => $inviteCode]);
    }

    /** @return array<string,mixed> */
    public function addParticipants(string $sessionId, string $groupId, array $participants): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/participants", [], ['participants' => $participants]);
    }

    /** @return array<string,mixed> */
    public function removeParticipants(string $sessionId, string $groupId, array $participants): array
    {
        return $this->http->request('DELETE', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/participants", [], ['participants' => $participants]);
    }

    /** @return array<string,mixed> */
    public function promoteParticipants(string $sessionId, string $groupId, array $participants): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/participants/promote", [], ['participants' => $participants]);
    }

    /** @return array<string,mixed> */
    public function demoteParticipants(string $sessionId, string $groupId, array $participants): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/participants/demote", [], ['participants' => $participants]);
    }

    /** @return array<string,mixed> */
    public function setSubject(string $sessionId, string $groupId, string $subject): array
    {
        return $this->http->request('PUT', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/subject", [], ['subject' => $subject]);
    }

    /** @return array<string,mixed> */
    public function setDescription(string $sessionId, string $groupId, string $description): array
    {
        return $this->http->request('PUT', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/description", [], ['description' => $description]);
    }

    /**
     * Get group settings. Only the settings the active engine supports are
     * present — any of {announce, locked, ephemeralSeconds} may be absent.
     *
     * @return array<string,mixed>
     */
    public function getGroupSettings(string $sessionId, string $groupId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/settings");
    }

    /**
     * Update group settings. At least one of {announce, locked,
     * ephemeralSeconds} is required (400 otherwise); ephemeralSeconds responds
     * 501 on engines without disappearing-message support (whatsapp-web.js).
     *
     * @param array<string,mixed> $settings
     * @return array<string,mixed>
     */
    public function updateGroupSettings(string $sessionId, string $groupId, array $settings): array
    {
        return $this->http->request('PUT', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/settings", [], $settings);
    }

    /** @return array<string,mixed> */
    public function leave(string $sessionId, string $groupId): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/leave");
    }

    /** @return array<string,mixed> */
    public function inviteCode(string $sessionId, string $groupId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/invite-code");
    }

    /** @return array<string,mixed> */
    public function revokeInviteCode(string $sessionId, string $groupId): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/groups/{$this->http->encodeSegment($groupId)}/invite-code/revoke");
    }
}
